tests/resources/test_dataset.py

```diff
@@ -82,6 +84,74 @@ def test_register_dataset(collection, session):
     assert name == dataset.name
 
 
+def test_register_dataset_with_idempotent_put(collection, session):
+    # Given
+    name = 'Test Dataset'
+    summary = 'testing summary'
+    description = 'testing description'
+    unique_name = 'foo'
+    session.set_response(DatasetDataFactory(name=name, summary=summary, description=description, unique_name=unique_name))
+
+    # When
+    session.use_idempotent_dataset_put = True
+    dataset = collection.register(DatasetFactory(name=name, summary=summary, description=description, unique_name=unique_name))
+
+    expected_call = FakeCall(
+        method='PUT',
+        path='projects/{}/datasets'.format(collection.project_id),
+        json={'name': name, 'summary': summary, 'description': description, 'unique_name': unique_name}
+    )
+    assert session.num_calls == 1
+    assert expected_call == session.last_call
+    assert name == dataset.name
```

Comment:
This has no failure modes being tested.

Reply:
With this method we test:
 - unique_name being None
 - no values being returned (this isn't an error, but is a branch)
 - > 1 results returned (that is an error)
 - exactly one value being returned

The actual endpoint can fail with an inappropriately formatted unique_name, but that's tested in the backend.

Not sure what else you're aiming to test? We already have coverage around session.get_resource, and collection.build. The purpose of this test is to ensure that with the given options that we invoke a PUT with the appropriate URL, and not the POST.

Reply:
> unique_name being None
You mean this scenario is covered by the existing `test_register_dataset`?

It's probably fine.  Part of the problem of trying to do PRs during meetings, probably.